lib/dcell/registries/mongodb_adapter.rb

```diff
@@ -15,71 +19,71 @@ def initialize(options)
         if options[:options]
           Mongoid.options = options[:options]
         end
+
+        @node_registry = Registry.new(DCellNode)
+        @global_registry = Registry.new(DCellGlobal)
       end
 
       class DCellNode
         include Mongoid::Document
 
         field :key, type: String
-        field :value, type: Hash
+        field :value, type: BSON::Binary
       end
 
       class DCellGlobal
         include Mongoid::Document
 
         field :key, type: String
-        field :value, type: Hash
+        field :value, type: BSON::Binary
       end
 
-      class Proxy
-        class << self
-          def set(storage, key, value)
-            entry = storage.find_or_create_by(key: key)
-            entry.value = {'v' => value}
-            entry.save!
-            value
-          end
+      class Registry
+        def initialize(storage)
+          @storage = storage
+        end
 
-          def get(storage, key)
-            first = storage.where(key: key).first
-            if first and first.value
-              return first.value['v']
-            end
-            nil
-          end
+        def _set(key, value, unique)
+          entry = @storage.find_or_create_by(key: key)
+          raise KeyExists if entry.value and unique
+          value = BSON::Binary.new(value.to_msgpack)
+          entry.value = value
+          entry.save!
+          value
+        end
 
-          def all(storage)
-            keys = []
-            storage.each do |entry|
-              keys << entry.key
-            end
-            keys
-          end
+        def set(key, value)
+          _set(key, value, false)
```

Comment:
It's switch to raise an exception if key-value tuple already exists. Anyway this should be trashed. I'll clean this up.